File: exam_LRX/py/work9.py
# 9.给出图像处理中空间域滤波的均值滤波、中值滤波和高斯滤波实例。
import cv2
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams

# 设置中文字体
rcParams['font.family'] = 'SimHei'  # 设置为黑体
rcParams['axes.unicode_minus'] = False  # 解决负号显示问题

# 均值滤波
def mean_filter(image):
    return cv2.blur(image, (11, 11))

# 中值滤波
def median_filter(image):
    return cv2.medianBlur(image, 15)

# 高斯滤波
def gaussian_filter(image):
    return cv2.GaussianBlur(image, (15, 15), 0)

# 显示所有滤波结果
def display_filters(image):
    # 进行滤波处理
    mean_img = mean_filter(image)
    median_img = median_filter(image)
    gaussian_img = gaussian_filter(image)

    # 创建一个包含四个子图的窗口
    fig, axes = plt.subplots(1, 4, figsize=(20, 5))

    # 显示原图
    axes[0].imshow(image, cmap='gray')
    axes[0].set_title('原始图像')
    axes[0].axis('off')  # 关闭坐标轴

    # 显示均值滤波后的图像
    axes[1].imshow(mean_img, cmap='gray')
    axes[1].set_title('均值滤波')
    axes[1].axis('off')

    # 显示中值滤波后的图像
    axes[2].imshow(median_img, cmap='gray')
    axes[2].set_title('中值滤波')
    axes[2].axis('off')

    # 显示高斯滤波后的图像
    axes[3].imshow(gaussian_img, cmap='gray')
    axes[3].set_title('高斯滤波')
    axes[3].axis('off')

    # 显示结果
    plt.tight_layout()
    plt.show()

# 主函数
def main():
    # 读取图像
    image_path = "../image/image9.png"  # 替换为你的图像路径
    image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)

    if image is None:
        print("图像读取失败！请检查路径是否正确。")
        return

    # 显示原图和滤波后的图像
    display_filters(image)

if __name__ == "__main__":
    main()

File: exam_LRX/cpp/work9.cpp
//
// Created by liryi on 24-11-22.
// 9.给出图像处理中空间域滤波的均值滤波、中值滤波和高斯滤波实例。
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

// 均值滤波
Mat meanFilter(const Mat& image) {
    Mat result;
    blur(image, result, Size(11, 11));  // 均值滤波
    return result;
}

// 中值滤波
Mat medianFilter(const Mat& image) {
    Mat result;
    medianBlur(image, result, 15);  // 中值滤波
    return result;
}

// 高斯滤波
Mat gaussianFilter(const Mat& image) {
    Mat result;
    GaussianBlur(image, result, Size(15, 15), 0);  // 高斯滤波
    return result;
}

// 在一个窗口中显示原图和滤波后的图像
void displayFilters(const Mat& originalImage, const Mat& meanImage, const Mat& medianImage, const Mat& gaussianImage) {
    // 创建一个大的显示矩阵
    int width = originalImage.cols;
    int height = originalImage.rows;
    Mat display = Mat::zeros(height, width * 4, originalImage.type());

    // 放置原始图像
    originalImage.copyTo(display(Rect(0, 0, width, height)));
    putText(display, "original image", Point(10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 放置均值滤波图像
    meanImage.copyTo(display(Rect(width, 0, width, height)));
    putText(display, "Mean filtering", Point(width + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 放置中值滤波图像
    medianImage.copyTo(display(Rect(width * 2, 0, width, height)));
    putText(display, "median filtering ", Point(width * 2 + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 放置高斯滤波图像
    gaussianImage.copyTo(display(Rect(width * 3, 0, width, height)));
    putText(display, "Gaussian filter", Point(width * 3 + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    // 显示结果
    namedWindow("Comparison of filtering effects", WINDOW_NORMAL);
    imshow("Comparison of filtering effects", display);
    waitKey(0);
}

int main() {
    // 读取图像
    string imagePath = "E:\\DL_class\\exam_LRX\\cpp\\images\\image9.png";  // 替换为你的图像路径
    Mat image = imread(imagePath, IMREAD_GRAYSCALE);

    if (image.empty()) {
        cout << "image error" << endl;
        return -1;
    }

    // 进行滤波
    Mat meanImage = meanFilter(image);
    Mat medianImage = medianFilter(image);
    Mat gaussianImage = gaussianFilter(image);

    // 显示所有结果
    displayFilters(image, meanImage, medianImage, gaussianImage);

    return 0;
}

